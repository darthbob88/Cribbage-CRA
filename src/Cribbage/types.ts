import { type } from "os";

interface GameState {
  players: Player[];
  // Torn between this, and making it a boolean field on the player.
  dealer: Player;
  currentHand: Hand;
}

//Yeah, I know this is bad and ambiguous 
interface Hand {
  cardsInHand: HandOfCards[];
  cardsInCrib: HandOfCards;
  playedCards: { player: Player; card: Card }[];
  turnUp: Card;
  runningTotal: number;
  whoseTurn: Player;
}

interface HandOfCards {
  player: Player;
  cards: Card[];
}
interface Player {
  id: string;
  score: number;
}

interface Card {
  suit: Suit;
  number: number;
}

enum Suit {
  Hearts = "HEARTS",
  Spades = "SPADES",
  Diamonds = "DIAMONDS",
  Clubs = "CLUBS"
}

export { Suit, Card, Hand, Player, HandOfCards }