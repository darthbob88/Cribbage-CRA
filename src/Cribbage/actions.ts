/* ACTION CREATOR FUNCTIONS

Put here the functions that return an action object that can be dispatched
HINT: Always use functions for consistency, don't export plain objects

*/

import * as types from "./types";

const quack = () => ({
    type: types.
});

const swim = (distance : number) => ({
    type: types.SWIM,
    payload: {
        distance
    }
});

export {
    swim,
    quack
};