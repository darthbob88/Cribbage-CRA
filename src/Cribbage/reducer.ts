import { fetch, addTask } from "domain-task";
import { Player, Card, GameState, Suit, Hand, Round, PHASE } from "./model";
import { Reducer, Action } from "redux";
import { typeKeys } from "./types";

//// Not sure if I'll actually use the full Redux machinery, or if I'll settle for just local state.
//// We are going FULL REDUX FTGE.
/* ACTIONS TO HANDLE
 * NEWGAME (player ID(s))
 * Initialize player name(s) to the IDs given, set their scores to 0, and dispatch NEWHAND
 * NEWHAND (newDealer)
 * Toggle which player is dealer and gets the crib
 * Initialize each player's hand to 6 cards, crib to 0, empty the list of played cards, and reset running total to 0.
 * SENDCARDSTOCRIB (player, cards)
 * Transfer cards from one player's hand to the crib
 * If both players have passed cards over, turn up deck for nobs and proceed to the play phase.
 * Maybe also do something with an ADDPOINTS action, that takes a player, action, and score?
 * PLAYCARD (player, card, (implicitly playedCards and running total))
 * Move card from player's hand to playedCards, add the value of that card to runningTotal, and toggle which player's turn it is.
 * If this would score points, dispatch ADDPOINTS as appropriate
 * If this would take the total to 31, reset total to 0, dispatch ADDPOINTS for 31 and continue play.
 * If neither player has cards left, dispatch "last card" and proceed to the counting-up phase.
 * PLAYGO (player)
 * Same as above, except no card transfer happens, we just toggle who goes next.
 * ADDPOINTS(player, points, string description?)
 * Just increment the number of points for <player>.
 * This might need us to store current and previous score, so we know where the back and front pegs should be.
 */

// -----------------
// ACTIONS - These are serializable (hence replayable) descriptions of state transitions.
// They do not themselves have any side-effects; they just describe something that is going to happen.

interface NewGameAction {
  // * NEWGAME(player ID(s))
  //* Initialize player name(s) to the IDs given, set their scores to 0, and dispatch NEWHAND
  type: typeKeys.newGame;
  playerNames: string[];
}

interface NewHandAction {
  // * NEWHAND(newDealer)
  //* Toggle which player is dealer and gets the crib
  //    * Initialize each player's hand to 6 cards, crib to 0, empty the list of played cards, and reset running total to 0.

  type: typeKeys.newHand;
  newDealer: Player;
  newHands: Hand[];
}

interface SendCardsToCribAction {
  /*
   * SENDCARDSTOCRIB (player, cards)
   * Transfer cards from one player's hand to the crib
   * If both players have passed cards over, turn up deck for nobs and proceed to the play phase.
   * Maybe also do something with an ADDPOINTS action, that takes a player, action, and score?
   */
  type: typeKeys.sendCardsToCrib;
  player: Player;
  cardsForCrib: Card[];
}
interface PlayCardAction {
  // * PLAYCARD(player, card, (implicitly playedCards and running total))
  //* Move card from player's hand to playedCards, add the value of that card to runningTotal, and toggle which player's turn it is.
  //    * If this would score points, dispatch ADDPOINTS as appropriate
  //    * If this would take the total to 31, reset total to 0, dispatch ADDPOINTS for 31 and continue play.
  //    * If neither player has cards left, dispatch "last card" and proceed to the counting - up phase.
  type: typeKeys.playCard;
  player: Player;
  card: Card;
}

interface PlayGoAction {
  //* PLAYGO(player)
  //* Same as above, except no card transfer happens, we just toggle who goes next.
  type: typeKeys.playGo;
  player: Player;
}

interface AddPointsAction {
  //* ADDPOINTS(player, points, string description ?)
  //* Just increment the number of points for <player>.
  //* This might need us to store current and previous score, so we know where the back and front pegs should be.
  type: typeKeys.addPoints;
  points: number;
  player: Player;
  description: string;
}

// Declare a 'discriminated union' type. This guarantees that all references to 'type' properties contain one of the
// declared type strings (and not any other arbitrary string).
type KnownAction =
  | NewGameAction
  | NewHandAction
  | SendCardsToCribAction
  | PlayCardAction
  | PlayGoAction
  | AddPointsAction;

// ----------------
// ACTION CREATORS - These are functions exposed to UI components that will trigger a state transition.
// They don't directly mutate state, but they can have external side-effects (such as loading data).

// export const actionCreators = {
//   requestWeatherForecasts: (
//     startDateIndex: number
//   ): AppThunkAction<KnownAction> => (dispatch, getState) => {
//     // Only load data if it's something we don't already have (and are not already loading)
//     if (startDateIndex !== getState().weatherForecasts.startDateIndex) {
//       let fetchTask = fetch(
//         `api/SampleData/WeatherForecasts?startDateIndex=${startDateIndex}`
//       )
//         .then(response => response.json() as Promise<WeatherForecast[]>)
//         .then(data => {
//           dispatch({
//             type: typeKeys.receiveForecast,
//             startDateIndex: startDateIndex,
//             forecasts: data
//           });
//         });

//       addTask(fetchTask); // Ensure server-side prerendering waits for this to complete
//       dispatch({
//         type: typeKeys.requestForecast,
//         startDateIndex: startDateIndex
//       });
//     }
//   }
// };

// ----------------
// REDUCER - For a given state and action, returns the new state. To support time travel, this must not mutate the old state.

export const initialState: GameState = {
  players: [],
  currentHand: {
    cardsInHand: [],
    crib: { player: "", cards: [] },
    dealer: "",
    phase: PHASE.DEAL,
    playedCards: [],
    turnUp: { suit: Suit.Spades, number: 1 },
    runningTotal: 0,
    whoseTurn: ""
  }
};

export const reducer: Reducer<GameState> = (
  state: GameState,
  incomingAction?: Action
) => {
  const action = incomingAction as KnownAction;
  switch (action.type) {
    case typeKeys.newGame:
      const namedPlayers = action.playerNames.map(name => (<Player>{ score: 0, id: name }));
      const dealer = action.playerNames[0];
      //TODO: This should probably be a function of some sort. Maybe an action creator, or just a "get a new hand of cards for each player"
      // Possibly even a "create a new hand of cards, including turnup".
      const newHand = <Partial<Round>>{
        playedCards: [],
        cardsInHand: [
          { player: action.playerNames[0], cards: [] },
          { player: action.playerNames[1], cards: [] }
        ],
        crib: { player: dealer, cards: [] },
        phase: PHASE.DEAL,
        runningTotal: 0,
        whoseTurn: dealer,
        dealer: dealer
      };
      return { players: namedPlayers, currentHand: newHand };
    case typeKeys.newHand:
      return initialState;
    case typeKeys.addPoints:
      return initialState;
    case typeKeys.playCard:
      return initialState;
    case typeKeys.playGo:
      return initialState;
    case typeKeys.sendCardsToCrib:
      return initialState;
    default:
      // The following line guarantees that every action in the KnownAction union has been covered by a case above
      const exhaustiveCheck: never = action;
  }

  return state || initialState;
};
