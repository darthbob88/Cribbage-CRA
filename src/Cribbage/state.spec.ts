import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import { reducer } from "./reducer";
import * as actions from "./actions";
import * as types from "./types";
import fetchMock from "fetch-mock";

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

// describe('async actions', () => {
//     afterEach(() => {
//         fetchMock.restore()
//     })

//     it('creates FETCH_TODOS_SUCCESS when fetching todos has been done', () => {
//         fetchMock.getOnce('/todos', {
//             body: { todos: ['do something'] },
//             headers: { 'content-type': 'application/json' }
//         })

//         const expectedActions = [
//             { type: types.FETCH_TODOS_REQUEST },
//             { type: types.FETCH_TODOS_SUCCESS, body: { todos: ['do something'] } }
//         ]
//         const store = mockStore({ todos: [] })

//         return store.dispatch(actions.fetchTodos()).then(() => {
//             // return of async actions
//             expect(store.getActions()).toEqual(expectedActions)
//         })
//     })
// })

//  * ACCEPTANCE REQUIREMENTS

//  * I should be able to tell the game my player name
describe("Player name", () => {
  it("should start with a blank name", () => {
    expect(reducer(undefined, {})).toEqual([
      { name: "", score: 0 },
      { name: "", score: 0 }
    ]);
  });
  it("should be able to update the name appropriately", () => {
    expect(
      reducer(undefined, { type: "SET_NAMES", names: ["Boris", "Natasha"] })
    ).toEqual([{ name: "Boris", score: 0 }, { name: "Natasha", score: 0 }]);
  });
});

//  * The game should then to deal 6 cards to each player
//   * And then each player should be able to pass exactly two cards to the crib.
//   * This should be the dealer's crib, not the off-dealer.

//  * After passing to the crib, we should commence to the PLAY phase.
//   * At this point, I should be able to play a card.
//     * In response, the game should
//       * properly increment the running total.
//       * register GO/31/Last card, and dispatch points/reset running total accordingly.
//       * register any scoring (15, pair, run, etc).
//       * Toggle whose turn it is.

//  * After all cards have been played, we proceed to the COUNT.
//   * This should go in the order dealer/off-dealer/dealer's crib.
//   * Scores should be dispatched per hand/crib, not in a single lump.

//  * After COUNT, the game should toggle who's dealer and deal a new hand.

//  * At some point, the game should register that somebody has won, and then whether the other player was (double)skunked.

//  * We should also handle pegging properly; jump the back peg over the front peg,
//  * and do it using a nice oblong board rather than just a couple straight lines
//   * Lower priority, though; for starters we'll just settle for "P1: 50 P2: 60"
