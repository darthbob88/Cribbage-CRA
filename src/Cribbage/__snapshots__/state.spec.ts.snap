// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Initializing the game should start with a blank name 1`] = `
Object {
  "currentHand": Object {
    "playedCards": Array [],
    "players": Array [],
    "runningTotal": 0,
    "turnUp": Object {
      "number": 1,
      "suit": "SPADES",
    },
    "whoseTurn": 1,
  },
  "dealer": Object {
    "cardsInCrib": Array [],
    "cardsInHand": Array [],
    "id": "",
    "score": 0,
  },
  "players": Array [],
}
`;
